# National Park Vacation Route Tool

## Data Source

Data was obtained from the National Park Service (https://www.nps.gov/index.htm) using their API (https://www.nps.gov/subjects/developer/api-documentation.htm).

## Data Preparation

### Import Libraries

In [ ]:
# JSON requests
import requests

# Data manipulation
import numpy as np
import pandas as pd

# Displaying plots and maps
import seaborn as sns
import matplotlib.pyplot as plt
import folium as fl

# Timing the algorithm
import time

# Display all fields
pd.set_option('display.max_columns', None)

# Ignore warnings
import warnings
warnings.filterwarnings('ignore')

### Import Raw Data from NPS API (external script)

In [ ]:
# (Moved to fetch_nps_data.py)
"""
# Available fields for each park

id
url
*fullName (used as name)
parkCode
name (short name)
*description
*designation
*latitude
*longitude
latLong (combined "lat:..., long:...")
*activities (array of objects with id and name)
*amenities
topics
*states
contacts (e.g., phone, email)
entranceFees (array)
entrancePasses
fees
directionsInfo
directionsUrl
operatingHours
addresses
images (photos array)
weatherInfo


API_KEY = "fpyJ9NycrgZX5mK8f0n90c4qXGPcYAsBPwt4BLJk"
url = "https://developer.nps.gov/api/v1/parks"

def fetch_all_parks(api_key):
    all_parks = []
    start = 0
    limit = 50

    while True:
        params = {
            "limit": limit,
            "start": start,
            "api_key": api_key
        }

        response = requests.get(url, params=params)
        response.raise_for_status()  # raise error for bad response
        data = response.json().get("data", [])

        if not data:
            break  # no more data

        all_parks.extend(data)
        start += limit  # go to next page

    return all_parks

# Fetch data
parks_raw = fetch_all_parks(API_KEY)

# Convert to DataFrame
records = []
for park in parks_raw:
    activity_list = park.get('activities', [])
    activity_names = [a.get('name', '') for a in activity_list]
    
    records.append({
        'name': park.get('fullName', ''),
        'latitude': park.get('latitude', ''),
        'longitude': park.get('longitude', ''),
        'designation': park.get('designation', ''),
        'states': park.get('states', ''),
        'description': park.get('description', ''),
        'activities': ', '.join(activity_names)
    })

parks = pd.DataFrame(records)

# Save as CSV
parks.to_csv("../data/nps_parks_with_activities.csv", index=False)
"""
# Run from fetch_nps_data.py
# %run ../scripts/fetch_nps_data.py

'\n# Available fields for each park\n\nid\nurl\n*fullName\nparkCode\nname\n*description\n*designation\n*latitude\n*longitude\nlatLong (combined "lat:..., long:...")\n*activities (array of objects with id and name)\ntopics\n*states\ncontacts (e.g., phone, email)\nentranceFees (array)\nentrancePasses\nfees\ndirectionsInfo\ndirectionsUrl\noperatingHours\naddresses\nimages (photos array)\nweatherInfo\n\n\nAPI_KEY = "fpyJ9NycrgZX5mK8f0n90c4qXGPcYAsBPwt4BLJk"\nurl = "https://developer.nps.gov/api/v1/parks"\n\ndef fetch_all_parks(api_key):\n    all_parks = []\n    start = 0\n    limit = 50\n\n    while True:\n        params = {\n            "limit": limit,\n            "start": start,\n            "api_key": api_key\n        }\n\n        response = requests.get(url, params=params)\n        response.raise_for_status()  # raise error for bad response\n        data = response.json().get("data", [])\n\n        if not data:\n            break  # no more data\n\n        all_parks.extend(data)\n    

### Import Generated CSV to Dataframe

In [3]:
parks = pd.read_csv('../data/nps_parks_with_activities.csv')

### Inspect Data

In [4]:
# Print dataframe head
print('First five rows of dataframe')
display(parks.head())
print()
    
# Print dataframe sample
print('Random five rows of dataframe')
display(parks.sample(5))
print()
    
# Check for missing values
print('Check for Missing Values')
print(parks.isna().sum())
print()

# Check data types
print('Check Data Types')
print(parks.info())
print()

# Check values for each column
print('Describe Dataframe')
print(parks.describe(include = 'all'))
print()
    
# Check for duplicates
print('Count of Duplicated Rows')
print(parks.duplicated().sum())
print()

# List unique activities (important for web interface)
print('Unique Activities')
print(parks['activities'].nunique())

First five rows of dataframe


name   latitude  longitude  \
0  Abraham Lincoln Birthplace National Historical...  37.585866 -85.673305   
1                               Acadia National Park  44.409286 -68.247501   
2                     Adams National Historical Park  42.255396 -71.011604   
3                African American Civil War Memorial  38.916600 -77.026000   
4            African Burial Ground National Monument  40.714527 -74.004474   

                designation states  \
0  National Historical Park     KY   
1             National Park     ME   
2  National Historical Park     MA   
3                       NaN     DC   
4         National Monument     NY   

                                         description  amenities  \
0  For over a century people from around the worl...        NaN   
1  Acadia National Park protects the natural beau...        NaN   
2  From the sweet little farm at the foot of Penn...        NaN   
3  Over 200,000 African-American soldiers and sai...        NaN   
4  The African Burial Ground is the oldest and la...        NaN   

                                          activities  
0  Astronomy, Stargazing, Food, Picnicking, Guide...  
1  Arts and Culture, Cultural Demonstrations, Ast...  
2  Guided Tours, Self-Guided Tours - Walking, Liv...  
3          Guided Tours, Self-Guided Tours - Walking  
4  Arts and Culture, Guided Tours, Junior Ranger ...


Random five rows of dataframe


name   latitude   longitude  \
303     Mormon Pioneer National Historic Trail  41.265032 -101.840838   
85      Chaco Culture National Historical Park  36.039634 -107.956008   
462    Women's Rights National Historical Park  42.908171  -76.816556   
471              Yucca House National Monument  37.247789 -108.686127   
435  Tuskegee Institute National Historic Site  32.429621  -85.704885   

                  designation          states  \
303   National Historic Trail  IL,IA,NE,UT,WY   
85   National Historical Park              NM   
462  National Historical Park              NY   
471         National Monument              CO   
435    National Historic Site              AL   

                                           description  amenities  \
303  Explore the Mormon Pioneer National Historic T...        NaN   
85   Explore the monumental structures and breathta...        NaN   
462  Women’s Rights National Historical Park tells ...        NaN   
471  Yucca House National Monument preserves a larg...        NaN   
435  In 1881, Booker T. Washington arrived in Alaba...        NaN   

                                            activities  
303  Auto and ATV, Scenic Driving, Guided Tours, Se...  
85   Astronomy, Camping, Guided Tours, Hiking, Juni...  
462  Guided Tours, Junior Ranger Program, Park Film...  
471          Guided Tours, Self-Guided Tours - Walking  
435  Guided Tours, Junior Ranger Program, Museum Ex...


Check for Missing Values
name             0
latitude         1
longitude        1
designation     35
states           0
description      0
amenities      474
activities      10
dtype: int64

Check Data Types
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 474 entries, 0 to 473
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         474 non-null    object 
 1   latitude     473 non-null    float64
 2   longitude    473 non-null    float64
 3   designation  439 non-null    object 
 4   states       474 non-null    object 
 5   description  474 non-null    object 
 6   amenities    0 non-null      float64
 7   activities   464 non-null    object 
dtypes: float64(3), object(5)
memory usage: 29.8+ KB
None

Describe Dataframe
                                                     name    latitude  \
count                                                 474  473.000000   
unique                                     

**Observations:**

- There are a few missing values - if we want to include those values on our map, these should be removed. These are latitude, longitude, designation, and activities
- Datatypes are all correct
- Amenities needs to be fixed (API issue?) or removed
- Activities between parks have a lot of overlap (431 unique vs. 474 parks!) and need to be streamlined/categorized
- The instructions want a maximum of 9 geographical locations, so we will have to trim a subset of this data to use
- We also have to remove National Parks not accessible by roads or extremely far away (e.g. Alaska, Hawaii, Puerto Rico, etc.)

### Clean Data
- Remove rows with missing coordinates
- Remove undesired parks (we have 474, probably way too many)
- Remove parks too far for a roadtrip or not accessible by land (Alaska, Hawaii, Puerto Rico, etc.) by bounding the latitude and longitude


In [5]:
# Drop rows with missing coordinates
parks = parks.dropna(subset=['latitude', 'longitude'])

### Map of All National Parks in the US

In [ ]:
fig = fl.Figure(width=1200, height=800)
m = Map(location=[39.8283, -98.5795], zoom_start=4)
fig.add_child(m)

for _, row in parks.iterrows():
    fl.Marker(
        location=[row['latitude'], row['longitude']],
        popup=row['name']
    ).add_to(m)

fig